# Сбор данных с Web-scraping и API для социально-научных исследований
---
## Подготовка к семинарам `7-8`. Работа с API ВКонтакте

---
*ФСН, ОП "Политология", 2023-2024 гг.*

Лика Капустина,

lkapustina@hse.ru

Наша следующая тема - **работа с API ВКонтакте**. Работа с частью API требует получения токена – специального ключа. Он может включать в себя информацию о вас, вашем профиле; токен может действовать постоянно или определенное количество времени.


**API ВКонтакте требует от вас создать специальный токен, чтобы получать данные и управлять определенными действиями с вашего аккаунта или с администрируемыми вами группами. В этом ноутбуке описано, как вы можете сгенерировать свой токен.**

<center><b><font size=6>Инструкция</font></b></center>

Инструкция основана на материале [devman.org](https://dvmn.org/encyclopedia/qna/63/kak-poluchit-token-polzovatelja-dlja-vkontakte/)


<font size=3><b>Шаг 1. Откройте раздел "Мои приложения" и создайте свое</b></font>

* Откройте страницу [Мои приложения](https://vk.com/apps?act=manage);
* Нажмите на кнопку "Создать";
* Придумайте название вашему приложению;
* Выберите платформу: **Standalone-приложение**;
* Остальные настройки не так важны (описание, тип). Вы можете не редактировать дефолтные значения.
* Кликните на кнопку "создать". 

<font size=3><b>Шаг 2. Достаньте client_id</b></font>

* Кликните на ваше приложение;
* Нажмите на кнопку "Редактировать";
* Нажмите на адресную строку. Скопируйте url;
* Вы получите url такого типа: ```https://vk.com/editapp?id=666``` где `id` равно `666`. Скопируйте значение id вашего приложения.

<font size=3><b>Шаг 3. Сгенерируйте ссылку из которой будете получать токен.</b></font>

* Проверьте [последнюю версию API VK](https://dev.vk.com/ru/reference/versions?ref=old_portal).
* Выберите значения для параметра `scope`. Все возможные указаны на соответствующей странице [API ВК](https://dev.vk.com/ru/reference/access-rights). Этот параметр отвечает за права доступа – то есть, за те действия, которые вы сможете совершать используя сгенерированную потом строку токена. 
    * Например: `'scope=friends'` позволит вам использовать дополнительные методы работы с друзьями (отправлять им сообщения, получать данные по ним и пр.;
    * Если вы хотите указать несколько `scope` (а я советую вам выбрать сразу несколько параметров), их нужно указать через запятую: `'friends,wall,status,wall'`
* Запустите код ниже или склейте строку самостоятельно.

In [21]:
api_version = input('Введите версию API ВК: ')      # последняя версия апи вк
client_id = input('Введите id вашего приложения: ') # client_id вашего приложения 

scope = 'friends,wall,status,wall'                  # настройки доступа

print('\nВаша ссылка для генерации API в ВК:\n')
print(f'https://oauth.vk.com/authorize?client_id={client_id}&display=page&scope={scope}&redirect_uri=https://oauth.vk.com/blank.html&response_type=token&v={api_version}&state=123456')

Введите версию API ВК: 5.1
Введите id вашего приложения: 666

Ваша ссылка для генерации API в ВК:

https://oauth.vk.com/authorize?client_id=666&display=page&scope=friends,wall,status,wall&redirect_uri=https://oauth.vk.com/blank.html&response_type=token&v=5.1&state=123456


<font size=3><b>Шаг 4. Работа с ссылкой выше.</b></font>

Запомните:<font color=red> НИКОМУ НЕЛЬЗЯ ПЕРЕДАВАТЬ ССЫЛКУ ВЫШЕ.</font> Передача этой ссылки грозит тем, что получивший её человек сможет с помощью методов API ВКонтакте управлять вашей страницей: например, рассылать какие-то сообщения вашим друзьям, удалить все записи на вашей странице и сделать другие неприятные вещи.

*Параметры `client_id` и токен, которые тут демонстрирую я, ненастоящие.*

**Сохраните эту ссылку в защищенную заметку или другое надежное место на вашем компьютере**.

Теперь, каждый раз когда вам нужно сгенерировать токен для работы с API ВКонтакте, вы будете переходить по этой ссылке. Ваш токен в дальнейшем будет использоваться для отправки запросов к API ВКонтакте.

**Перейдите по ссылке**

Вы должны увидеть следующую запись: 

```Пожалуйста, не копируйте данные из адресной строки для сторонних сайтов. Таким образом Вы можете потерять доступ к Вашему аккаунту.```

Кликните на поле с адресом страницы. Теперь там должна появиться следующая строка:

```https://oauth.vk.com/blank.html#access_token=vk1.a.h1t15y9MSMQtVe0jlX4jeSA&expires_in=86400&user_id=000666```, где:
* `access_token` – сам токен доступа. Эта строка будет выглядеть сильно больше, но для этого примера я сократила реальный токен;
* `expires_in=86400` это число секунд, через которые истечет срок действия этого токена. 86 тысяч 400 секунд это 24 часа;
* `user_id=000666` это ваш id ВКонтакте.

Протестируйте, работает ли ваш токен:

In [4]:
token = input('Введите ваш токен ВКонтакте: ') # сохраним токен (начиная с vk1.a.)

Введите ваш токен ВКонтакте: vk1.a.h1t15y9MSMQtVe0jlX4jeSA&expires_in=86400&user_id=000666


In [5]:
# импортируем библиотеки
import requests
import pandas as pd
import numpy

In [18]:
# отправим запрос к вашей странице
# воспользуемся подготовленной мной функцией
def get_user_data(user_ids, method='users.get'):
    parameters = 'fields=sex,about,screen_name,bdate,city,country,home_town'
    url = 'https://api.vk.com/method/' + method + '?' + parameters + '&v=' + version + '&access_token=' + token + '&user_ids=' + user_ids
    response = requests.get(url)
    users_data = response.json()
    
    # Данные
    user_ids = [users_data['response'][i]['id'] for i in range(len(users_data['response']))]
    first_names = [users_data['response'][i]['first_name'] for i in range(len(users_data['response']))]
    last_names = [users_data['response'][i]['last_name'] for i in range(len(users_data['response']))]
    users_sex = [{0: 'Н', 1: 'Ж', 2: 'М'}[users_data['response'][i]['sex']] for i in range(len(users_data['response']))]
    is_closed_accounts = [users_data['response'][i]['is_closed'] for i in range(len(users_data['response']))]        
    
    # создаем датафрейм по пользователям
    df_users = pd.DataFrame({'id': user_ids, 'first_name': first_names, 'last_name': last_names, 
                             'sex': users_sex, 'is_closed': is_closed_accounts})
    df_users['first_name'] = df_users['first_name'] + ' ' + df_users['last_name']
    df_users.drop('last_name', axis=1, inplace=True)
    df_users = df_users.rename(columns={'first_name': 'full_name'})
    df_users = df_users.fillna('')
    # возвращаем датафрейм
    return df_users

В результате отправки запроса вы должны увидеть краткую информацию о себе. Проверьте, что совпадает ваше имя (`full_name`), указанный пол - `sex` – (нет, в примере не ошибка), и статус – `is_closed` (`True`/`False`).

In [19]:
your_user_id = token.split('user_id=')[1] # тут получаем id нашей страницы (не client_id)
version = api_version # сохраняем в переменную version версию АПИ ВКонтакте

get_user_data(user_ids=f'{your_user_id}') 

id       full_name sex  is_closed
0  441721976  Lika Kapustina   М       True

Если все совпадает – **поздравляю, вы можете пользоваться вашим токеном!**

Если нет – пройдитесь по инструкции еще раз. Если у вас не получилось, то напишите об этом мне в чате курса.